### ISSUE TITLE
Syntax for expanded source capability

### ISSUE THREAD

**Issue by paulromano** [mit-crpg/openmc#530]
I wanted to get some opinions on extending our XML syntax for a significantly improved capability to specify distributions for starting sources. There are a few things I'm planning on implementing:

1. Ability to specify multiple source distributions
2. Ability to decouple x, y, and z components of spatial distribution
3. Ability to specify distributions for mu and phi for the angular distribution
4. An abstract univariate probability distribution with a variety of extensions that can be used for any single variable (x-coordinate, mu for angle, energy)
5. Tabular probability distributions (piecewise continuous as well as discrete)

Together, these capabilities will give the user much more flexibility in specifying source distributions. I already have 3, 4, and 5 implemented and now I'm starting to think about 1 and 2. Namely, **I want to know if there is any preference between these two possible options for having multiple source distributions**
``` XML
<source probability="0.4">
  <space ... /><angle ... /><energy ... />
</source>
<source probability="0.6">
  <space ... /><angle ... /><energy ... />
</source>
```
OR
```XML
<source>
  <distribution probability="0.4">
    <space ... /><angle ... /><energy ... />
  </distribution>
  <distribution probability="0.6">
     <space ... /><angle ... /><energy ... />
  </distribution>
</source>
```
The first one would be sort-of backwards compatible, but backwards compatibility is going to be broken by other changes anyway.

For the univariate probability distributions, I am planning on having something like the following:
```XML
<var1 type="tabular" interpolation="linear" parameters="-1. 0. 1. 0.5 0.5 0.5" />
<var2 type="discrete" parameters="10. 20. 50. 0.3 0.4 0.3" />
<var3 type="uniform" parameters="-3.1415926 3.1415926" />
<var4 type="watt" parameters="0.988 2.249" />
```
So for example, if you wanted to manually specify an isotropic distribution, it would look like:
```XML
<source>
  <angle type="mu-phi">
    <mu type="uniform" parameters="-1. 1." />
    <phi type="uniform" parameters="0. 6.2831853" />
  </angle>
</source>
```
for which there would be a shorthand
```XML
<source>
   <angle type="isotropic" />
</source.
```

Angular distributions will also support a "reference_uvw" attribute that indicates what direction the polar angle is relative to. This means a monodirectional distribution would look as follows:
```XML
<source>
  <angle type="monodirectional" reference_uvw="1.0 0.0 0.0" />
</source>
```

The way the Python API handles source is pretty ugly right now, so I'm going to extend that as well. I have some thoughts about how I'd like to do that, but I'm open to whatever thoughts and ideas others may have.

@smharper I think you were the one who mentioned last week that you'd like to see the capability to have multiple source distributions, so I'm hoping you'll opine on this issue.

**Comment by smharper** [comment]
I'd be happy to opine on this issue.

I slightly prefer the first of those two syntaxes for multiple sources.  I cannot think of an attribute that we would want two different sources to share.  And if they share no attributes, then I think it makes more sense conceptually if they are separate entities.

Also, I think we should (eventually if not now) have the ability to rejection sample sources for materials and cells.  I think we should also have a tally filter for the source a particle was born from.
```xml
<source id="1" probability="0.4">
  <space ... /><angle ... /><energy ... />
  <cells>1 2 3</cells>
  <materials>1 2 3</materials>
</source>
<source id="2" ... />

<tally id="1">
  <scores>flux</scores>
  <filter type="source" bins="1 2" />
</tally>
```

**Comment by wbinventor** [comment]
I've never used the fixed source feature in OpenMC, but I can see that the
changes proposed here would greatly improve the current functionality. I'm
inclined to prefer the second of the two proposed syntax options. Also, I
find the idea of separate XML tags for different distributions to be a bit
odd.

Based on the examples, a source will be a multinomial distribution over
nested user-defined distributions over angle, space and energy variables.
This leads me to wonder if the XML syntax would be more expressive if it
reflected the fact that what we are dealing are distributions over
distributions.
For example, perhaps the Python API might be extended with a new
"Distribution" class with "var," "type," "parameters," etc. attributes. The
"var" could *either* take on a value like "mu," "phi," etc. *or* a "nested"
distribution. Of course the XML syntax could then be developed to reflect
the OO implementation in the API. I recognize my approach here reflects the
more philosophical question - should one start with XML syntax or the API
when determining user interface design?

Sorry for the lack of specificity. While I'm on travel the next few days
I'll give some thought to how a more general distributions-of-distributions
approach might be realized.

I'd be happy to opine on this issue.

I slightly prefer the first of those two syntaxes for multiple sources. I
cannot think of an attribute that we would want two different sources to
share. And if they share no attributes, then I think it makes more sense
conceptually if they are separate entities.

Also, I think we should (eventually if not now) have the ability to
rejection sample sources for materials and cells. I think we should also
have a tally filter for the source a particle was born from.

<source id="1" probability="0.4">
  <space ... /><angle ... /><energy ... />
  <cells>1 2 3</cells>
  <materials>1 2 3</materials>
</source>
<source id="2" ... />

<tally id="1">
  <scores>flux</scores>
  <filter type="source" bins="1 2" />
</tally>

—
Reply to this email directly or view it on GitHub
<https://github.com/mit-crpg/openmc/issues/530#issuecomment-165529521>.

**Comment by nelsonag** [comment]
Seems like this ticket was closed by #556, but I cant be confident in that since I wasnt really involved in the review of that PR and havent used many 'fancy' source definition tricks.  Can this issue be closed?

**Comment by paulromano** [comment]
Yes, this can be considered closed.
